### PR TITLE
Improve Data.Set module docs.

### PR DESCRIPTION
**Pull request by m-renaud**
This improves the module level docs for Data.Set following the form of the recent improvements to Data.Sequence. This is for the most part a re-organization of the existing content with a few minor changes so should hopefully be uncontroversial.

Rendered docs can be found [here](http://www.m-renaud.com/haddocks/containers-0.5.10.2/Data-Set.html).

Partially addresses #495.

### CODE REVIEW

File: Data/Set.hs [hunk 1]
```diff
@@ -13,14 +13,48 @@
 -- Maintainer  :  libraries@haskell.org
 -- Portability :  portable
 --
--- An efficient implementation of sets.
+--
+-- = Sets
+--
+-- The @'Set' e@ type represents a set of /unique/, /ordered/ elements of type
+-- @e@.
```
Comment (treeowl) [inline Data/Set.hs]: What are ordered elements? What is a unique element? I think perhaps

    -- The @'Set' e@ type represents a set of elements of type @e@. Most operations
    -- require that @e@ be an instance of the 'Ord' class.

File: Data/Set.hs [hunk 2]
```diff
@@ -13,14 +13,48 @@
...
+--
+-- For a walkthrough of the most commonly used functions see the
+-- <https://haskell-containers.readthedocs.io/en/latest/set.html sets introduction>.
+--
+-- Note that the implementation is /left-biased/ -- the elements of a
+-- first argument are always preferred to the second, for example in
+-- 'union' or 'insert'.  Of course, left-biasing can only be observed
+-- when equality is an equivalence relation instead of structural
+-- equality.
```
Comment (treeowl) [inline Data/Set.hs]: See my note on `Map`. Also, rather than "left-biasing", how about "this bias"?

File: Data/Set.hs [hunk 3]
```diff
@@ -13,14 +13,48 @@
...
+--
+-- == Strictness properties
+--
+-- This module satisfies the following strictness property:
+--
+-- * Key arguments are evaluated to WHNF
+--
+-- Here are some examples that illustrate the property:
+--
+-- > delete undefined s  ==  undefined
```
Comment (treeowl) [inline Data/Set.hs]: This demonstrates nothing. How about

```haskell
-- > insert undefined xs == undefined
-- > singleton undefined = undefined
```

File: Data/Set.hs (hunk 1 above)
Comment (m-renaud) [reply Data/Set.hs]: Done.

File: Data/Set.hs (hunk 3 above)
Comment (m-renaud) [reply Data/Set.hs]: Done.

File: Data/Set.hs (hunk 2 above)
Comment (m-renaud) [reply Data/Set.hs]: Done.

**Comment by treeowl** [comment]
I don't see the update.

**Comment by m-renaud** [comment]
Oops, accidentally sent the review comments before pushing the commit. I thought I had "started a review" which I was going to finish after the commit went up. I'll have it up shortly, about to walk to work.

**Comment by m-renaud** [comment]
Changes pushed, Haddocks haven't been pushed again because its a pain to get both changes built simultaneously because they're in different branches.

File: Data/Set.hs [hunk 4]
```diff
@@ -13,14 +13,38 @@
 -- Maintainer  :  libraries@haskell.org
 -- Portability :  portable
 --
--- An efficient implementation of sets.
+--
+-- = Finite Sets
+--
+-- The @'Set' e@ type represents a set of elements of type @e@. A 'Set' is
+-- strict in the element type and most operations require that @e@ be an
```
Comment (treeowl) [inline Data/Set.hs]: Surely you mean strict in its *elements*, not their *type*. I think you should move the strictness to a new sentence after the one about `Ord`.

Comment (m-renaud) [reply Data/Set.hs]: Oops, yup that what I meant.

Comment (m-renaud) [reply Data/Set.hs]: Maybe this should just be a bullet list, it may read better since both the sentences are so short. WDYT?

**Comment by treeowl** [comment]
I don't think bullets will help.

**Comment by m-renaud** [comment]
How about: "Most operations require that @e@ be an instance of the 'Ord' class, and all operations are strict in the element."

**Comment by treeowl** [comment]
Technically, I suppose that's true. But it's much more intuitively helpful,
I believe, to think of Set as a strict data structure.

**Comment by m-renaud** [comment]
Gotcha, I'll leave it as is. Are there any other changes you'd like made or can this be merged?

**Pull request merged**